crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4180,6 +4224,104 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn except_handler() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            import builtins
+
+            try:
+                x
+            except NameError as e:
+                pass
+            except (TypeError, builtins.AttributeError) as f:
+                pass
+            ",
+        )?;
+
+        assert_public_ty(&db, "src/a.py", "e", "NameError");
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // TODO should be `TypeError | AttributeError` (needs support for tuple types) --Alex
+        assert_public_ty(&db, "src/a.py", "f", "Unknown");
+
+        Ok(())
+    }
+
+    #[test]
+    fn unknown_type_in_except_handler_does_not_cause_spurious_diagnostic() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            from nonexistent_module import foo
+
+            try:
+                x
+            except foo as e:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(
+            &db,
+            "src/a.py",
+            &["Cannot resolve import 'nonexistent_module'."],
+        );
+        assert_public_ty(&db, "src/a.py", "foo", "Unknown");
+        assert_public_ty(&db, "src/a.py", "e", "Unknown");
+
+        Ok(())
+    }
+
+    #[test]
+    fn dynamic_exception_handler() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            EXCEPTIONS = (AttributeError, TypeError)
+
+            try:
+                x
+            except EXCEPTIONS as e:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
+
+        // TODO: Should be `AttributeError | TypeError` --Alex
+        assert_public_ty(&db, "src/a.py", "e", "Unknown");
+
+        Ok(())
+    }
+
+    #[test]
+    fn exception_handler_with_invalid_syntax() -> anyhow::Result<()> {
+        let mut db = setup_db();
+
+        db.write_dedented(
+            "src/a.py",
+            "
+            try:
+                x
+            except as e:
+                pass
+            ",
+        )?;
+
+        assert_file_diagnostics(&db, "src/a.py", &[]);
```

Comment:
I checked manually that we do emit a diagnostic informing the user about invalid syntax in this snippet (though the message we give currently isn't great; I wasn't sure at first whether it was a diagnostic or an internal panic!).

It would be nice if `assert_file_diagnostics()` also included diagnostics we emit due to syntax errors so that we got a complete picture of the diagnostics we were emitting. But considering that we're about to implement a whole new set of testing infrastructure, I don't know if that's worth spendingn time on.